src/EventListener/Api/PublishableEventListener.php

```diff
@@ -0,0 +1,125 @@
+            $this->handlePUTRequest($classMetadata, $configuration, $data, $request);
+        }
+    }
+
+    private function handlePOSTRequest(ClassMetadataInfo $classMetadata, Publishable $configuration, object $data): void
+    {
+        // It's not possible for a user to define a resource as draft from another
+        $classMetadata->setFieldValue($data, $configuration->associationName, null);
+
+        // User doesn't have draft access: force publication date
+        if (!$this->publishableHelper->isGranted()) {
+            $classMetadata->setFieldValue($data, $configuration->fieldName, new \DateTimeImmutable());
+        }
+    }
+
+    private function handlePUTRequest(ClassMetadataInfo $classMetadata, Publishable $configuration, object $data, Request $request): void
+    {
+        $changeSet = $this->getEntityManager($data)->getUnitOfWork()->getEntityChangeSet($data);
+
+        // It's not possible to change the publishedResource property
+        if (isset($changeSet[$configuration->associationName])) {
+            $classMetadata->setFieldValue($data, $configuration->associationName, $changeSet[$configuration->associationName][0]);
+        }
+
+        // User doesn't have draft access: cannot change the publication date
+        if (!$this->publishableHelper->isGranted()) {
+            if (isset($changeSet[$configuration->fieldName])) {
+                $classMetadata->setFieldValue($data, $configuration->fieldName, $changeSet[$configuration->fieldName][0]);
+            }
+
+            // Nothing to do here anymore for user without draft access
+            return;
+        }
+
+        // User requested for original object
+        if (true === $request->query->getBoolean('published', false)) {
+            // User cannot change the publication date of the original resource
+            if ($changeSet[$configuration->fieldName]) {
+                throw new BadRequestHttpException('You cannot change the publication date of a published resource.');
+            }
+
+            // User wants to update the original object: nothing to do here anymore
+            return;
+        }
+
+        // Resource is a draft of another resource: nothing to do here anymore
+        if (null !== $classMetadata->getFieldValue($data, $configuration->associationName)) {
+            return;
+        }
+
+        // Any field has been modified: create or update draft
+        $draft = $this->getEntityManager($data)->getRepository($this->getObjectClass($data))->findOneBy([
+            $configuration->associationName => $data,
+        ]);
+        if (!$draft) {
+            $draft = clone $data;
+            // Reset draft identifier(s)
+            $classMetadata->setIdentifierValues($draft, array_combine($classMetadata->getIdentifierFieldNames(), array_fill(0, \count($classMetadata->getIdentifierFieldNames()), null)));
+
+            $this->getEntityManager($draft)->persist($draft);
```

Comment:
Flush is already done by API Platform. Here we just switch the objects, and to prevent any update in the database by a global flush, it's better to refresh the original object